database/sqlite/sqlite_db_migration.c

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "sqlite_db_migration.h"
+
+static int do_migration_noop(sqlite3 *database, const char *name)
+{
+    UNUSED(database);
+    UNUSED(name);
+    info("Running database migration %s", name);
+    return 0;
+}
+
+static struct database_func_migration_list {
+    char *name;
+    int (*func)(sqlite3 *database, const char *name);
+} migration_action[] = {
+    {.name = "v0 to v1",  .func = do_migration_noop},
+    // the terminator of this array
+    {.name = NULL, .func = NULL}
+};
+
+
+static int perform_database_migration_cb(void *data, int argc, char **argv, char **column)
+{
+    int *status = data;
+    UNUSED(argc);
+    UNUSED(column);
+    *status = str2uint32_t(argv[0]);
+    return 0;
+}
+
+int perform_database_migration(sqlite3 *database, int target_version)
+{
+    int user_version = 0;
+    char *err_msg = NULL;
+
+    int rc = sqlite3_exec(database, "PRAGMA user_version;", perform_database_migration_cb, (void *) &user_version, &err_msg);
+    if (rc != SQLITE_OK) {
+        info("Error checking the database version; %s", err_msg);
+        sqlite3_free(err_msg);
+    }
```

Comment:
I know this most likely will not fail, but if it does, should we attempt to run all migration functions from 0 (default user_version) to target? I.e. most likely the user will already be on target version, should we default maybe user version to target version and skip the upgrade functions?

Reply:
It is not clear if its possible to run all migrations from 0 to version X. 

It depends how the migration functions will be written to make it safe. For example dropping a column from a table will require to create a new temp table copy data and drop the temp table. 

The "regular" config sequence will attempt to create always the new table with the correct number of columns. In this case the migration will fail if it runs again (because the old column does not exist). 

In another scenario it may be possible without a lot of effort. So dropping a table as part of moving from v1 to v2 --  will be a DROP IF EXISTS command which is fine to run again. 

Even in the case of "dropped" columns, it is possible to check the schema and decide what to do